tmt/steps/prepare/artifact/providers/repository_provider.py

```diff
@@ -0,0 +1,115 @@
+"""
+This file provides classes for handling repository file artifacts in the tmt framework.
+It includes the `RepositoryFileInfo` class, which encapsulates metadata about a repository
+file (e.g., URL and filename), and the `RepositoryFileProvider` class, which manages the
+downloading of `.repo` files to a guest system.
+"""
+
+from collections.abc import Iterator
+from shlex import quote
+from urllib.parse import urlparse
+
+import tmt.log
+import tmt.utils
+from tmt.container import container
+from tmt.steps.prepare.artifact.providers import (
+    ArtifactProvider,
+    DownloadError,
+)
+from tmt.steps.prepare.artifact.providers.info import ArtifactInfo
+from tmt.steps.provision import Guest
+from tmt.utils import GeneralError, Path, ShellScript
+
+
+@container
+class RepositoryFileInfo(ArtifactInfo):
+    """
+    A single repository file channel.
+
+    Encapsulates information about a repository file, including its URL and
+    derived identifier (filename). It validates that the URL uses HTTP/HTTPS
+    and points to a file with a ``.repo`` extension.
+    """
+
+    _raw_artifact: str
+
+    def __post_init__(self) -> None:
+        """
+        Validate the repository file URL after initialization.
+
+        Raises:
+            ValueError: If the URL scheme is not HTTP/HTTPS or if the file
+                        does not have a ``.repo`` extension.
+        """
+        parsed_url = urlparse(self._raw_artifact)
+        if parsed_url.scheme not in ('http', 'https'):
+            raise ValueError(
+                f"Invalid repository URL: {self._raw_artifact}. Only HTTP(S) is supported."
+            )
+        path = tmt.utils.Path(parsed_url.path)
+        if path.suffix != '.repo':
+            raise ValueError(f"URL must point to a .repo file, got: {path.name}")
+        self._id = path.name
+
+    @property
+    def id(self) -> str:
+        """The filename of the repository file"""
+        return self._id
+
+    @property
+    def location(self) -> str:
+        """The location (URL) of the repository file"""
+        return self._raw_artifact
+
+
+class RepositoryFileProvider(ArtifactProvider[RepositoryFileInfo]):
```

Comment:
Note that this should be converted to `ArtifactProvider[RpmArtifactInfo]` because we will be listing RPMs. It is unclear without the user part of it since the whole `download_artifacts` needs equivalent adjustments, but at least put a TODO note about that.

Reply:
Added Todo in [commit](https://github.com/teemtee/tmt/pull/4018/commits/fd34ba18d1c80ac7be30d79d7a69f559ab9396f9).